### PR TITLE
fix: update rule types

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

- **ADDED** the missing type definitions for the following rules:
   - `no-useless-assignment`
   - `no-object-constructor`
   - `no-new-native-nonconstructor`
   
- **UPDATED** the outdated type definitions for the following rules:
   - `no-fallthrough` - Added `reportUnusedFallthroughComment` option type.
   - `require-unicode-regexp` -  Added `requireFlag` option type. 
   - `no-new-object` -  Marked as deprecated in favor of `no-object-constructor`
   - `no-new-symbol` - Marked as deprecated in favor of `no-new-native-nonconstructor`
   - `no-useless-computed-key` -  Added `enforceForClassMembers` option 
   - `func-style` - Added `overrides` option type.
   - `lines-between-class-members` - Added `enforce` option type.
   - `no-unused-vars` - Added `ignoreClassWithStaticInitBlock` & `reportUsedIgnorePattern` option type & updated default value for the `caughtErrors` option.
   
- **REMOVED** the type definitions for the following rules that no longer exist:  
   - `require-jsdoc`
   - `valid-jsdoc`

#### Is there anything you'd like reviewers to focus on?
No

### CODE REVIEW

**Comment by nzakas** [comment]
@snitin315 it would be helpful to have a more specific PR title to understand what it is you're working on.

File: lib/types/rules/ecmascript-6.d.ts
Comment (snitin315) [inline lib/types/rules/ecmascript-6.d.ts:225]:
```suggestion
     *
```

**Review by nzakas — APPROVED** [review]
LGTM. Would like another review before merging.

File: lib/types/rules/deprecated.d.ts
Comment (JoshuaKGoldberg) [inline lib/types/rules/deprecated.d.ts:207]:
```suggestion
     * @deprecated since 8.27.0, use [`no-new-native-nonconstructor`](https://eslint.org/docs/rules/no-new-native-nonconstructor) instead.
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
🙌!

I read through the corresponding rules on the docs site and confirmed everything matches up... and am now also looking forward to having these be auto-generated (#18912). 😂

File: lib/types/rules/ecmascript-6.d.ts
Comment (Tanujkanti4441) [inline lib/types/rules/ecmascript-6.d.ts:224]:
```suggestion
     * Rule to disallow `new` operator with global non-constructor functions
```

**Review by nzakas — APPROVED** [review]
Re-approved. Waiting for @JoshuaKGoldberg to also re-approve.

**Comment by snitin315** [comment]
Friendly ping @JoshuaKGoldberg

**Review by JoshuaKGoldberg — APPROVED** [review]
Sorry for the delay! This looks great to me!

**Review by amareshsm — APPROVED** [review]
LGTM

**Pull request merged**
